### ISSUE TITLE
macOS 10.15: slow command autocompletion

### ISSUE THREAD

**Issue by dimaqq** [fish-shell/fish-shell#6270]
Ever since Catalina update, I experience ~10s delays every now and again (not often, but typically at least once after boot and maybe less than once an hour).

For example, I have `poetry` installed ( https://github.com/sdispater/poetry ) the point is only that it installs something "executable".

When I type `poe` in the shell, fish tries to show dark gray autocomplete, in my case typically `poetry shell`. Usually it's instantaneous, but sometimes it takes 10s during which I cannot type anything... or rather I can type by terminal is not updated.

My setup is `iTerm2` + `fish` from pkg installer + usual dev things, git, yarn, npm, python3.8, poetry, vim, etc.

```
> fish --version
fish, version 3.0.2
```

I'm not actually entirely sure what causes the sudden lag and I don't know if I can in fact blame it on `fish`, but I'd love to figure that out :)

When I've upgraded to Catalina, I had to manually "allow access" for iterm2 to Downloads, Documents, etc... I wonder if that's connected?

Alternatively, could there be something with `hostname` occasionally lagging?

**Comment by krobelus** [comment]
If you can reproduce this with fish launched as `fish --profile my-profile-file`, you can look at the breakdown in `my-profile-file` to see what is taking so much time.

If you use the command `hostname` you can try the variable `$hostname` instead but that should not make a difference.

**Comment by Fingel** [comment]
I believe I have the same issue. After upgrading to Catalina, I noticed completions became very slow. Today I decided to look into it. I launched fish with `fish -d5`. I noticed anytime I tried to tab-complete something, I'd get this output:

```
<4> fish: select_try on fd 7
<4> fish: select_try hit timeout
<4> fish: select_try: no fds returned valid data within the timeout
```
Repeated like 50 times. Whether or not that is related to the next issue, I'm not sure.

I did a little more digging and discovered that the output of the `whatis` and `apropos` commands contained a lot of useless garbage:

```
$ whatis man
hugo-gen-man(1)          - Generate man pages for the Hugo CLI
groff_man(7)             - groff `man' macros to support generation of man pages
groffer(1)               - display groff files and man~pages on X and tty
man(1)                   - format and display the on-line manual pages
man.conf(5)              - configuration data for man
zshall(1)                - the Z shell meta-man page
xml2man(1)               - MPGL to mdoc (man page) translator
makewhatis: /usr/lib/./libgutenprint.2.dylib: No such file or directory
makewhatis: /usr/lib/libsasl2.2.0.1.dylib: Not a directory
makewhatis: /usr/lib/libldap.dylib: Not a directory
makewhatis: /usr/lib/libsqlite3.0.dylib: Not a directory
makewhatis: /usr/lib/libcom_err.dylib: Not a directory
```
Followed by at least another 50 lines of the "Not a directory" messages. Same thing with `apropos` The commands take like ~500ms to complete.

I'm pretty sure fish uses either apropos or whatis for it's completions. So I suspect if this is the issue, it's not fish's fault.

To add to this, I thought perhaps rebuilding the `whatis` database would remove the errors (those libraries maybe were upgraded with Catalina?). However, it doesn't seem possible:

```
$ sudo /usr/libexec/makewhatis
Password:
makewhatis: /usr/share/man/whatis.tmp: Read-only file system
```

Which make sense because now Catalina has /usr on a read-only file system. 

I've posted a question about this on [Ask Different](https://apple.stackexchange.com/questions/374025/errors-from-whatis-command-unable-to-rebuild-database-with-makewhatis) as well as the [Apple support forums](https://discussions.apple.com/thread/250806874)

If my suspicions are correct, this seems like a problem with Catalina. Could they have not realized that the `makewhatis` command needs access to /usr/share/man ? It's possible.

Or I'm way off and the slowdown is coming from something else. Either way, it's bothering me, and anything I can do to help I will.

**Comment by Fingel** [comment]
I did launch a fish profile instance as suggested by @krobelus , and looking through it found this:

```
429736	429736	-> apropos $argv 2>/dev/null | awk -v FS=" +- +" '{
		split($1, names, ", ");
		for (name in names)
			if (names[name] ~ /^'"$argv"'.* *\([18]\)/ ) {
				sub( "( |\t)*\\\([18]\\\)", "", names[name] );
				sub( " \\\[.*\\\]", "", names[name] );
				print names[name] "\t" $2;
			}
	}'
```

The 429736  429736 seem like large times compared to everything else. But I'm not totally sure what I'm looking at.

Full profile is here: https://gist.github.com/Fingel/7c2292a54977c06321f7daf72d81bca6

**Comment by faho** [comment]
Yeah, that's fish starting `apropos` to provide command descriptions, and that's it taking ~430ms (the time is in microseconds).

**Comment by Fingel** [comment]
Ok, so that lines up with my observations that running apropos and whatis outside of fish returns a bunch of garbage and takes about that long to complete.

Is there any way to disable command descriptions?

**Comment by faho** [comment]
>Is there any way to disable command descriptions?

Not without patching fish, no.

**Comment by Fingel** [comment]
Alright, well my suspicion then is that this isn't a bug in fish, but a bug in Catalina. Maybe I'll get some answers in other the other venues I posted about rebuilding the whatis database. I'm not hopeful though.

**Comment by faho** [comment]
>Alright, well my suspicion then is that this isn't a bug in fish, but a bug in Catalina. 

Unfortunately that doesn't really matter. As much as I'd love to just tell Apple to fix their stuff, or people to move away from it until it is fixed, realistically that's not going to work.

So we'll probably have to find a solution, and most likely a proper one would take a while.

Soo... we might want to introduce a flag to disable this? Maybe at compile-time, so the homebrew people can enable it on Catalina?

**Comment by faho** [comment]
As a workaround, replace `__fish_describe_command` with:

```fish
function __fish_describe_command; end
```

**Comment by faho** [comment]
Compare #5825.

**Comment by Fingel** [comment]
Thanks you @faho ! That really helped.

For those that aren't super familiar with fish that might find this issue, the easiest way to edit a function in fish is to use funced:

    funced __fish_describe_command

And then just remove or comment out the code there.

Then save it:

    funcsave __fish_describe_command

That'll save your definition locally so you're not overriding anything in fish itself. Great shell.

**Comment by dimaqq** [comment]
Those with enough stack overflow mojo could perhaps offer a bounty on @Fingel 's question at:
https://apple.stackexchange.com/questions/374025/errors-from-whatis-command-unable-to-rebuild-database-with-makewhatis 🙏

**Comment by krobelus** [comment]
Does it make sense to add a workaround in fish 3.1?

**Comment by ridiculousfish** [comment]
If someone can suggest an easy workaround then I think it makes sense to take it; nothing comes to my mind though. Apple really needs to fix this.

**Comment by faho** [comment]
So the workaround is to just do nothing in `__fish_describe_command`. The tricky part is detecting when to apply it.

We could, when loading the function, check the macOS version, but that wouldn't reenable it when/if Apple applies a fix.

**Comment by ridiculousfish** [comment]
The Mac's `apropos` [is pretty short](https://gist.github.com/ridiculousfish/8e01a725f7645f694001f3f211954e34), it would not be crazy to just bypass it entirely and write our own that doesn't run `makewhatis`.

**Comment by floam** [comment]
Yeah. It's just grepping a text file. At one point in ancient history fish apparently read the file directly in the C code to improve the speed over shelling out.

**Comment by dimaqq** [comment]
> Yeah. It's just grepping a text file. At one point in ancient history fish apparently read the file directly in the C code to improve the speed over shelling out.

`mmap` like Silver Searcher `ag` does would rock :)

**Comment by faho** [comment]
This should have *some* workaround for 3.1.0.

If nobody comes up with anything better, I'm just gonna commit

```fish
if test (uname) = Darwin
    function __fish_describe_command; end
else
   # Use the actual function
end
```

**Comment by floam** [comment]
I'll commit a better workaround.

**Comment by faho** [comment]
Okay, can someone try the following in functions/__fish_describe_command.fish:

```fish
# macOS 10.15 "Catalina" has *issues*
if test (uname) = Darwin
    if command -sq defaults
        set -l darwin_version (defaults read loginwindow SystemVersionStampAsString | string split .)
        if test "$darwin_version[1]" = 10 -a "$darwin_version[2]" = 15
            function __fish_describe_command; end
            exit
        end
    end
end

# And here the real definition
```

This should disable __fish_describe_command on Catalina and leave it as-is everywhere else. We can make the version more specific if need be.

**Comment by floam** [comment]
Reading the version using `defaults` like that is going to be pretty slow, like 300ms.

`sw_vers -productVersion` outputs `10.15.3` and took 14ms. Or `uname -r` (Catalina is Darwin 19 - on 10.15.3 that outputs `19.3.0`), it takes 4ms.

**Comment by faho** [comment]
Okay, so 

```fish
if test (uname) = Darwin
    set -l darwin_version (uname -r | string split .)
        # macOS 15 is Darwin 19
        if test "$darwin_version[1]" = 19 -a "$darwin_version[2]" -le 3
            function __fish_describe_command; end
            exit
        end
    end
end
```

Assuming that it'll be fixed in 10.15.4. If we don't think they will, that second $darwin_version should just be removed.

**Comment by ridiculousfish** [comment]
I would be surprised if this got fixed in a point release, I will remove the second `$darwin_version` check.

**Comment by faho** [comment]
@ridiculousfish I thought similarly, which is why I commented it out.

**Comment by floam** [comment]
I'm going to commit something that instead of putting in an empty function, on macOS we'll grep any `whatis` files in the users' man paths, just like the apropos shell script does. With the key difference being we won't try update the whatis databases if any whatis files are absent.

This will have the effect of being able to describe some commands - commands that came with Xcode, stuff installed in /usr/local/... and whenever Apple does fix it and /usr/share/man/whatis exists again it'll pick those up without us putting out an update.

**Comment by krobelus** [comment]
Yeah, we can override both `apropos` and `whatis` (bound to <kbd>Alt</kbd><kbd>W</kbd>) with functions.

**Comment by suizman** [comment]
As an alternative workaround you can use the binaries supplied with [Plan9port](https://github.com/9fans/plan9port) which are faster than the MacOS ones.

**Comment by 100ideas** [comment]
here's my solution for now, based on the [solution linked above (stack overflow)](https://apple.stackexchange.com/questions/374025/errors-from-whatis-command-unable-to-rebuild-database-with-makewhatis), which involved making a patched copy of apple's `apropos` script in which `/usr/lib` is removed from the list of locations apropos searches.

Then I placed my patched copy of `apropos` in a location included at the front of my `$fish_user_paths`

I wasn't sure if my local copy would be sourced instead of the system copy in internal fish scripts, so I checked where `__fish_describe_command` was getting it and it seemed to be using mine. N

ot sure if this will always true, i.e. not true when in non-tty shell mode no? thoughts appreciated.

My patched copy of apropos is below, see inline comments for specifics.

---

```sh
#!/bin/sh

############################################
# 2020-02-12: patched apropos b/c apple broke it, making fish completions slow
#
# this file should be stored in dir included at front of $fish_user_paths or $PATH to override
# system apropos
#
# issue: slow tab command completion in fish
#  - https://github.com/fish-shell/fish-shell/issues/6270
# solution:
#  - https://apple.stackexchange.com/questions/374025/errors-from-whatis-command-unable-to-rebuild-database-with-makewhatis
# rough procedure:
#
#   ```sh
#   $ mkdir -p ~/workarounds
#   $ sed -e 66s@/usr/lib@@ /usr/bin/apropos > ~/workarounds/apropos.macos_10.15.1
#   $ diff /usr/bin/apropos ~/workarounds/apropos.macos_10.15.1
#   66c66
#   <     for d in /var/cache/man $manpath /usr/lib
#   ---
#   >     for d in /var/cache/man $manpath
#   $ chmod +x ~/workarounds/apropos.macos_10.15.1
#   $ mv ~/workarounds/apropos.macos_10.15.1 ~/dev/bin
#
#   # note: `~/dev/bin` is at the front of my PATH; I keep local scripts there
#   $ echo $fish_user_paths
#     /Users/100ideas/dev/bin /usr/local/opt/curl/bin /usr/local/opt/sqlite/bin
#   $ which apropos
#     /Users/100ideas/dev/bin/apropos
#   ```
#
#  to verify fish autocomplete uses the patched apropos, I added
#   ```echo "location of apropos" (which apropos)``` to end of
#  `__fish_describe_command`
#
#   ```sh
#   $ _fish_describe_command you
#   youtube-dl	download videos from youtube.com or other video platforms
#   location of apropos /Users/100ideas/dev/bin/apropos
#   ```
#
# use this file to override native apropos
############################################



# apropos -- search the whatis database for keywords.
# whatis  -- idem, but match only commands (as whole words).
#
# Copyright (c) 1990, 1991, John W. Eaton.
# Copyright (c) 1994-1999, Andries E. Brouwer.
#
# You may distribute under the terms of the GNU General Public
# License as specified in the README file that comes with the man
# distribution.
#
# apropos/whatis-1.5m aeb 2003-08-01 (from man-1.6g)
#
# keep old PATH - 000323 - Bryan Henderson
# also look in /var/cache/man - 030801 - aeb

program=`basename $0`

# When man pages in your favorite locale look to grep like binary files
# (and you use GNU grep) you may want to add the 'a' option to *grepopt1.
aproposgrepopt1='i'
aproposgrepopt2=''
whatisgrepopt1='iw'
whatisgrepopt2=''
grepopt1=$aproposgrepopt1
grepopt2=$aproposgrepopt2

if [ $# = 0 ]
then
    echo "usage: $program keyword ..."
    exit 1
fi

manpath=`man --path | tr : '\040'`

if [ "$manpath" = "" ]
then
    echo "$program: manpath is null"
    exit 1
fi

args=
for arg in $*; do
    case $arg in
        --version|-V|-v)
	    echo "$program from man-1.6g"
	    exit 0
	    ;;
	--help|-h)
            echo "usage: $program keyword ..."
	    exit 0
	    ;;
	-*)
	    echo "$program: $arg: unknown option"
	    exit 1
	    ;;
	*)
	    args="$args $arg"
    esac
done

while [ "$1" != "" ]
do
    found=0
    for d in /var/cache/man $manpath
    do
        if [ -f $d/whatis ]
        then
            if grep -"$grepopt1" "$grepopt2""$1" $d/whatis
            then
                found=1
# Some people are satisfied with a single occurrence
# But it is better to give all
#               break
            fi
        else
            if [ -d $d ] && /usr/libexec/makewhatis.local "-o /dev/fd/1" $d | grep -"$grepopt1" "$grepopt2""$1"
            then
                found=1
# Some people are satisfied with a single occurrence
# But it is better to give all
#               break
            fi
        fi
    done

    if [ $found = 0 ]
    then
        echo "$1: nothing appropriate"
    fi

    shift
done | eval ${PAGER:-more -E}

exit
```

**Comment by Fingel** [comment]
The underlying issue with `whatis` and `apropos` seems to have been silently fixed in MacOS 10.15.4

**Comment by zanchey** [comment]
Could I ask someone on 10.15.4, who *doesn't* have a special apropos command installed, to check that the following patch doesn't reproduce the original behaviour?

```diff
diff --git a/share/functions/__fish_complete_man.fish b/share/functions/__fish_complete_man.fish
index 3654ef303..318e9a211 100644
--- a/share/functions/__fish_complete_man.fish
+++ b/share/functions/__fish_complete_man.fish
@@ -6,9 +6,8 @@
 # `apropos` with their own, which presumably doesn't have the same problem.
 if test (uname) = Darwin
     set -l darwin_version (uname -r | string split .)
-    # macOS 15 is Darwin 19, this is an issue at least up to 10.15.3.
-    # If this is fixed in later versions uncomment the second check.
-    if test "$darwin_version[1]" = 19 # -a "$darwin_version[2]" -le 3
+    # macOS 15 is Darwin 19, this is an issue up to and including 10.15.3.
+    if test "$darwin_version[1]" = 19 -a "$darwin_version[2]" -le 3
         set -l apropos (command -s apropos)
         if test "$apropos" = /usr/bin/apropos
             function __fish_complete_man
diff --git a/share/functions/__fish_describe_command.fish b/share/functions/__fish_describe_command.fish
index 433a2166d..f13fad472 100644
--- a/share/functions/__fish_describe_command.fish
+++ b/share/functions/__fish_describe_command.fish
@@ -9,9 +9,8 @@
 # So we disable this entirely in that case.
 if test (uname) = Darwin
     set -l darwin_version (uname -r | string split .)
-    # macOS 15 is Darwin 19, this is an issue at least up to 10.15.3.
-    # If this is fixed in later versions uncomment the second check.
-    if test "$darwin_version[1]" = 19 # -a "$darwin_version[2]" -le 3
+    # macOS 15 is Darwin 19, this is an issue up to and including 10.15.3.
+    if test "$darwin_version[1]" = 19 -a "$darwin_version[2]" -le 3
         function __fish_describe_command
         end
         # (remember: exit when `source`ing only exits the file, not the shell)
```

**Comment by ridiculousfish** [comment]
`apropos` is indeed much faster on 10.15.4. However for whatever reason, on my Mac tab completing commands has a generic description:

    grep  (Executable, 46kB)

whereas on Linux I get:

    grep (Print lines that match patterns)

I need to figure out why, `__fish_describe_command` seems to work correctly on my Mac.

**Comment by ridiculousfish** [comment]
Ok, the prefix form like `__fish_describe_command gre` now returns empty on both Mac and Linux; bisects to 535845861a1e4a63131790a1e4bf04c0b68dfa61 . This has broken tab completion descriptions.

**Comment by ridiculousfish** [comment]
@zanchey your change works well for me on 10.15.4

**Comment by muzaffar-omer** [comment]
I faced similar issue in Catalina 10.15.2, and it was fixed when I updated to the latest `fish` version `3.1.2`

Thank you all for following up the issue and finding a fix 🙏

**Comment by faho** [comment]
Important note: This bug was reintroduced in macOS 11, in which the workaround no longer applies (because we had version-gated it).

That's been fixed in #7365 with a better workaround that will not be version-gated.

Please don't file any more issues unless you can reproduce the problem *with* 7365 applied, which will be released in fish 3.2.0.
